dali/pipeline/executor/executor2/op_task.cc

```diff
@@ -0,0 +1,366 @@
+    std::vector<OutputDesc> output_descs;
+    output_descs.reserve(nout);
+    // If Setup returns true, we must resize the outputs;
+    if (node_->op->Setup(output_descs, ws)) {
+      assert(output_descs.size() == static_cast<size_t>(nout));
+      for (int i = 0; i < nout; i++) {
+        if (ws.OutputIsType<CPUBackend>(i)) {
+          ws.Output<CPUBackend>(i).Resize(output_descs[i].shape, output_descs[i].type);
+        } else if (ws.OutputIsType<GPUBackend>(i)) {
+          ws.Output<GPUBackend>(i).Resize(output_descs[i].shape, output_descs[i].type);
+        } else {
+          assert(!"Unreachable code - unknown backend.");
+        }
+      }
+    }
+  }
+}
+
+void OpTask::RunOp() {
+  if (!skip_) {
+    DomainTimeRange tr("[DALI][Executor] Run");
+    node_->op->Run(*ws_);
+    ResetInputLayouts();
+    PropagateSourceInfo(*ws_);
+  }
+  if (auto cpt = ws_->GetIterationData()->checkpoint) {
+    node_->op->SaveState(cpt->GetOpCheckpoint(node_->instance_name), ws_->output_order());
+  }
+  if (ws_->has_event() && ws_->has_stream())
+    CUDA_CALL(cudaEventRecord(ws_->event(), ws_->stream()));
+}
+
+void OpTask::SetWorkspaceInputs() {
+  int ti = 0;
+  assert(ws_->NumInput() + ws_->NumArgumentInput() == static_cast<int>(node_->inputs.size()));
+  auto order = ws_->output_order();
+  std::unordered_set<cudaEvent_t> events(ws_->NumInput() + ws_->NumArgumentInput());
+  for (int i = 0; i < ws_->NumInput(); i++, ti++) {
+    if (ws_->InputIsType<CPUBackend>(i)) {
+      auto inp = TaskInput<CPUBackend>(ti);
+      if (inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    } else {
+      assert(ws_->InputIsType<GPUBackend>(i));
+      auto inp = TaskInput<GPUBackend>(ti);
+      // If the output order of the operator is `host` then we don't wait for GPU
+      // inputs - they can't be accessed directly on host and the operator will
+      // have to issue some sort of synchronization if and when necessary.
+      // This optimization is essential to avoid oversynchronization
+      // when the operator needs to access the metadata only (e.g. getting the shape).
+      if (order.is_device() /*see comment above */ && inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    }
+  }
+
+  for (int i = 0; i < ws_->NumArgumentInput(); i++, ti++) {
+    auto &inp = TaskInput<CPUBackend>(ti);
+    if (inp.event)
+      events.insert(inp.event);
+    ws_->SetArgumentInput(i, inp.data);
+  }
+
+  if (task_->NumInputs() > ti) {  // Artificial error input
+    // TODO(michalz): Add early failure mechanism to `dali::tasking` that makes the successors
+    //                of a task that threw an exception fail. Currently the error disappears,
+    //                so we have to add an artificial input which propagates the exception.
+    task_->GetInputValue<void>(ti++);
+  }
+
+  for (auto e : events)
+    ws_->output_order().wait(e);
```

Comment:
Is it just a case of weird naming, and output_order is the "run order" of this operator, so we wait for inputs to be ready to use when we run the operator?

Reply:
Yes, there's no distinct `run_order`.